File: algorithms/2_medium/M/02429_minimize_xor/solution.c
int minimizeXor(const int num1_int, const int num2_int) {
    assert(num1_int >= 1 && num2_int >= 1);
    const unsigned int num1 = (unsigned int)num1_int;
    const unsigned int num2 = (unsigned int)num2_int;
    size_t ones = __builtin_popcount(num2);
    unsigned int res = 0;

    for (int i = 31; i >= 0; i -= 1) {
        if (ones == 0) { break; }

        const unsigned int mask = (1u << (unsigned int)i);
        if ((mask & num1) != 0) {
            res |= mask;
            ones -= 1;
        }
    }

    for (unsigned int i = 0; i < 32; i += 1) {
        if (ones == 0) { break; }

        const unsigned int mask = (1u << i);
        if ((mask & num1) == 0) {
            res |= mask;
            ones -= 1;
        }
    }

    return (int)res;
}

File: algorithms/2_medium/M/02429_minimize_xor/solution.rs
impl Solution {
    pub fn minimize_xor(num1: i32, num2: i32) -> i32 {
        assert!(num1 >= 1 && num2 >= 1);
        let (num1, num2) = (num1 as u32, num2 as u32);
        let mut ones = num2.count_ones();
        let mut res: u32 = 0;

        for i in (0..32_u32).rev() {
            if ones == 0 { break; }

            let mask: u32 = (1 << i);
            if (mask & num1) != 0 {
                res |= mask;
                ones -= 1;
            }
        }

        for i in 0..32_u32 {
            if ones == 0 { break; }

            let mask: u32 = (1 << i);
            if (mask & num1) == 0 {
                res |= mask;
                ones -= 1;
            }
        }

        res as i32
    }
}
